components/omega/OmegaBuild.cmake

```diff
@@ -77,61 +82,261 @@ macro(preset)
+
+  message(STATUS "OMEGA_C_COMPILER = ${OMEGA_C_COMPILER}")
+  message(STATUS "OMEGA_CXX_COMPILER = ${OMEGA_CXX_COMPILER}")
+  message(STATUS "OMEGA_Fortran_COMPILER = ${OMEGA_Fortran_COMPILER}")
+
+  # detect OMEGA_ARCH if not provided
+  if(NOT OMEGA_ARCH)
+
+    if(USE_CUDA)
+      set(OMEGA_ARCH "CUDA")
+
+    elseif(USE_HIP)
+      set(OMEGA_ARCH "HIP")
+
     else()
-      find_program(_OMEGA_Fortran_COMPILER ${MPIFC})
+
+      execute_process(
+        COMMAND ${OMEGA_CXX_COMPILER} --version
+        RESULT_VARIABLE _CXX_VER_RESULT
+        OUTPUT_VARIABLE _CXX_VER_OUTPUT)
+
+      if (_CXX_VER_RESULT EQUAL 0)
+
+        string(REGEX MATCH "HIP|hip"       _HIP_CHECK "${_CXX_VER_OUTPUT}")
+        string(REGEX MATCH "AMD|amd"       _AMD_CHECK "${_CXX_VER_OUTPUT}")
+        string(REGEX MATCH "NVCC|nvcc"     _NVCC_CHECK "${_CXX_VER_OUTPUT}")
+        string(REGEX MATCH "NVIDIA|nvidia" _NVIDIA_CHECK "${_CXX_VER_OUTPUT}")
+
+        if(_HIP_CHECK AND _AMD_CHECK)
+          set(OMEGA_ARCH "HIP")
+
+        elseif(_NVCC_CHECK AND _NVIDIA_CHECK)
+          set(OMEGA_ARCH "CUDA")
+
+        else()
+          set(OMEGA_ARCH "OPENMP")
+
+        endif()
+      else()
+        set(OMEGA_ARCH "OPENMP")
+
+      endif()
     endif()
   endif()
 
-  set(OMEGA_C_COMPILER ${_OMEGA_C_COMPILER})
-  set(CMAKE_C_COMPILER ${OMEGA_C_COMPILER})
+  message(STATUS "OMEGA_ARCH = ${OMEGA_ARCH}")
 
-  set(OMEGA_CXX_COMPILER ${_OMEGA_CXX_COMPILER})
-  set(CMAKE_CXX_COMPILER ${OMEGA_CXX_COMPILER})
+  # create a profile script
+  set(_ProfileScript ${OMEGA_BUILD_DIR}/omega_profile.sh)
+  file(WRITE ${_ProfileScript}  "#!/usr/bin/env bash\n\n")
+  file(APPEND ${_ProfileScript} "source ./omega_env.sh\n\n")
+  file(APPEND ${_ProfileScript} "# modify 'OUTFILE' with a path in that the profiler can\n")
+  file(APPEND ${_ProfileScript} "# create files such as a path in a scratch file system.\n")
 
-  set(OMEGA_Fortran_COMPILER ${_OMEGA_Fortran_COMPILER})
+  # set C and Fortran compilers *before* calling CMake project()
+  set(CMAKE_C_COMPILER ${OMEGA_C_COMPILER})
   set(CMAKE_Fortran_COMPILER ${OMEGA_Fortran_COMPILER})
 
-  message(STATUS "OMEGA_C_COMPILER = ${OMEGA_C_COMPILER}")
-  message(STATUS "OMEGA_CXX_COMPILER = ${OMEGA_CXX_COMPILER}")
-  message(STATUS "OMEGA_Fortran_COMPILER = ${OMEGA_Fortran_COMPILER}")
-  message(STATUS "MPI_EXEC = ${MPI_EXEC}")
+# TODO: do we want to use these variables?
+#  # Set compiler and linker flags
+#  if (CXXFLAGS)
+#    separate_arguments(_CXXFLAGS NATIVE_COMMAND ${CXXFLAGS})
+#    list(APPEND OMEGA_CXX_FLAGS ${_CXXFLAGS})
+#  endif()
+#
+#  if (LDFLAGS)
+#    separate_arguments(_LDFLAGS NATIVE_COMMAND ${LDFLAGS})
+#    list(APPEND OMEGA_LINK_OPTIONS ${_LDFLAGS})
+#  endif()
+#
+#  if (SLIBS)
+#    separate_arguments(_SLIBS NATIVE_COMMAND ${SLIBS})
+#    list(APPEND OMEGA_LINK_OPTIONS ${_SLIBS})
+#  endif()
+
+  # set CXX compiler *before* calling CMake project()
+  if(OMEGA_ARCH STREQUAL "CUDA")
+
+    if(NOT OMEGA_CUDA_COMPILER)
```

Comment:
Do you really need to manually handle nvcc_wrapper? Kokkos is already elegantly handling this, via cake target properties. Everything that links against kokkos automatically gets compiled with nvcc wrapper.

In particular, you can use mpicxx (or whatever your mpi compiler is) as CMAKE_CXX_COMPILER. Kokkos will make sure that nvcc wrapper is used behind the scene when appropriate.

Reply:
@bartgol, I manually used 'nvcc_wrapper' because I was unable to make it work with other approaches, including the mpicxx approach on Perlmutter GPU with the Nvidia HPC compiler.

Is 'CMAKE_CXX_COMPILER' the only variable I need to set, or are there other CMake variables I should modify? If there are any examples I could refer to, could you please share them with me? I'd like to try them on PM-GPU.

Reply:
Usually you only need to set CMAKE_CXX_COMPILER. Kokkos will then inject the `launch_compiler` script for the compilation step of _anything_ that links against kokkos (*). In the `launch_compiler` script, kokkos invokes `CMAKE_CXX_COMPILER` for host side compilation, and nvcc_wrapper for device compilation.

What happens if you just use `CMAKE_CXX_COMPILER=mpicxx` (or CC, or whatever is the mpi wrapper name)?

(*) Provided that it's C++ code. Also, if you ever link with PRIVATE, the kokkos compilation hack does not propagate further downstream.

Reply:
Thanks for the explanation. I think that this approach looks more robust than manually using nvcc_wrapper, and I will try it. Thanks a lot!